### PR TITLE
WIP: Move code to ES6

### PR DESCRIPTION
**Pull request by ChrisAlderson**
This is still a work in progress and shouldn't be merged for now, but I wanted to show you how the butter-provider would look in a ES6 way.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,206 +1,229 @@
-var assign = Object.assign || require('es6-object-assign').assign
-var memoize = require('memoizee')
+'use strict'
 
-var parseArgForType = function (type, arg) {
+const memoize = require('memoizee')
```
Comment (xaiki) [inline index.js]: why not use the `import` syntax here ?

File: index.js [hunk 2]
```diff
@@ -1,206 +1,229 @@
...
+  constructor(args = defaultArgs, config = defaultConfig) {
+    this.args = args
+    this.config = config
+
+    const memopts = args.memops
+
+    this.args = Object.assign(
+      this.args,
+      this._processArgs(this.args)
+    )
```
Comment (xaiki) [inline index.js]: we set `this.args` twice here.

File: index.js [hunk 3]
```diff
@@ -1,206 +1,229 @@
...
+    this.detail = this._makeCached(this.detail, memopts)
+  }
+
+  _makeCached(method, memopts) {
+    const self = this
+    const memoizedMethod = memoize(method, memopts)
+
+    return function() {
+      // XXX: Should be replaced with spread operator if possible.
+      return memoizedMethod.apply(this, arguments)
```
Comment (xaiki) [inline index.js]: we should also use the `() => ()` syntax

File: index.js [hunk 4]
```diff
@@ -1,206 +1,229 @@
...
 
-    // XXX:reimplement querystring.parse to not escape
-    var args = {}
-    tokenize[1] && tokenize[1].split('&').map(function (v){
-        var m = v.split('=')
-        args[m[0]]= parseArgForType(config.args[m[0]], m[1])
+    // XXX: This really isn't readable code.
+    tokenize[1] && tokenize[1].split('&').map(v => {
+      const pair = v.split('=')
+      const [ key, value ] = pair
```
Comment (xaiki) [inline index.js]: `const [ key, value ] = v.split('=')`

File: index.js [hunk 5]
```diff
@@ -1,206 +1,229 @@
...
-    DEFAULT: '0',
-    LOW: '480p',
-    MEDIUM: '720p',
-    HIGH: '1080p'
-}
+    return Promise.resolve(oldData)
+  }
+
+  toString(arg) {
+    return JSON.stringify(this)
```
Comment (xaiki) [inline index.js]: `this` looks wrong, shouldn't be `arg` ?

File: index.js [hunk 6]
```diff
@@ -1,206 +1,229 @@
...
+  NULL: null
 }
 
-Provider.prototype.toString = function (arg) {
-    return JSON.stringify(this)
+Provider.QualityType = {
+  DEFAULT: '0',
+  LOW: '480p',
+  MEDIUM: '720p',
+  HIGH: '1080p'
```
Comment (xaiki) [inline index.js]: probably needs a 'null' arg too.

File: tests/basic.js [hunk 7]
```diff
@@ -1,182 +1,184 @@
...
-    return new P(testArgs)
+  return new P(testArgs)
 }
 
-function isInValues(element, set) {
-    return Object.keys(set).reduce(function (a, c) {
-        return a || (element === set[c])
-    })
+function isInValues (element, set) {
+  return Object.keys(set).reduce((a, c) => a || (element === set[c]))
```
Comment (xaiki) [inline tests/basic.js]: i thought you needed `()` for it to return

File: index.js (hunk 1 above)
Comment (ChrisAlderson) [inline index.js]: For the `import` syntax you need to use `babel`. I could add `babel` if you want and also make the static properties defined at the end (like ArgType, ItemType etc) as class properties.

File: index.js (hunk 3 above)
Comment (ChrisAlderson) [inline index.js]: Arrow functions mess with the lexical scope: https://babeljs.io/learn-es2015/#ecmascript-2015-features-arrows-and-lexical-this

File: index.js (hunk 5 above)
Comment (ChrisAlderson) [inline index.js]: I don't know you tell me: https://github.com/butterproviders/butter-provider/blob/master/index.js#L202

Comment (xaiki) [inline index.js]: i guess this is unused ? looks wrong...

File: package.json [hunk 8]
```diff
@@ -17,7 +17,6 @@
   },
   "homepage": "https://github.com/butterproviders/butter-provider#readme",
   "dependencies": {
-    "debug": "^2.6.8",
```
Comment (xaiki) [inline package.json]: we cactually want moar debug =)

File: index.js [hunk 9]
```diff
@@ -132,7 +123,10 @@ class Provider {
 
     const uniqueId = this.config.uniqueId
     return this.fetch()
-      .then(({ results }) => this._randomArray(results))
+      .then(({ results }) => {
+        const random = Math.floor(Math.random() * results.length)
+        return results[random]
```
Comment (xaiki) [inline index.js]: that's what randomArray does, why reimplement it ?

Comment (ChrisAlderson) [inline index.js]: `randomArray` was only used once in the `random` method. So unless you also want to use it elsewhere I thought it was best to just implement it in the `random` method.

Comment (xaiki) [inline index.js]: i guess it's good as a utility function, but it's not currently used.

File: index.js (hunk 1 above)
Comment (xaiki) [inline index.js]: ok

File: index.js (hunk 3 above)
Comment (xaiki) [inline index.js]: ok, i guess this should always be us though.

**Pull request merged**